File: crates/psychevo-gateway/tests/fixtures/fake_acp_cancel.py
#!/usr/bin/env python3
import json
import sys

log_path = sys.argv[1]
prompt_id = None


def send(value):
    print(json.dumps(value), flush=True)


def record(method):
    with open(log_path, "a", encoding="utf-8") as log_file:
        log_file.write(method + "\n")


for line in sys.stdin:
    if not line.strip():
        continue
    message = json.loads(line)
    method = message.get("method")
    mid = message.get("id")
    record(method or "response")
    if method == "initialize":
        send(
            {
                "jsonrpc": "2.0",
                "id": mid,
                "result": {"protocolVersion": 1, "agentCapabilities": {}},
            }
        )
    elif method == "session/new":
        send(
            {
                "jsonrpc": "2.0",
                "id": mid,
                "result": {"sessionId": "native-cancel"},
            }
        )
    elif method == "session/prompt":
        prompt_id = mid
    elif method == "session/cancel" and prompt_id is not None:
        send(
            {
                "jsonrpc": "2.0",
                "id": prompt_id,
                "result": {"stopReason": "end_turn"},
            }
        )
        prompt_id = None

File: crates/psychevo-gateway/tests/fixtures/fake_acp_cancel.js
#!/usr/bin/env node
const fs = require("node:fs");
const readline = require("node:readline");
const logPath = process.argv[2];
let promptId = null;

function send(value) { process.stdout.write(`${JSON.stringify(value)}\n`); }

readline.createInterface({ input: process.stdin, crlfDelay: Infinity }).on("line", (line) => {
  if (!line.trim()) return;
  const message = JSON.parse(line);
  const method = message.method;
  fs.appendFileSync(logPath, `${method || "response"}\n`, "utf8");
  if (method === "initialize") send({ jsonrpc: "2.0", id: message.id, result: { protocolVersion: 1, agentCapabilities: {} } });
  else if (method === "session/new") send({ jsonrpc: "2.0", id: message.id, result: { sessionId: "native-cancel" } });
  else if (method === "session/prompt") promptId = message.id;
  else if (method === "session/cancel" && promptId != null) {
    send({ jsonrpc: "2.0", id: promptId, result: { stopReason: "end_turn" } });
    promptId = null;
  }
});
